### PR TITLE
[Behat][API] Browsing channels using their default locales

### PR DESCRIPTION
**Pull request by coldic3**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets |
| License         | MIT

### CODE REVIEW

File: features/locale/handling_different_locales_on_multiple_channels.feature [hunk 1]
```diff
@@ -1,33 +1,33 @@
...
         And I should not be able to shop using the "angielski (Stany Zjednoczone)" locale
 
-    @ui
+    @ui @api
     Scenario: Browsing channels using their default locales
-        When I browse the "Web" channel
+        Given I browse the "Web" channel
         And I start browsing the "Mobile" channel
-        Then I should shop using the "polski (Polska)" locale
+        Then I should shop using the "Polish (Poland)" locale
```
Comment (coldic3) [inline features/locale/handling_different_locales_on_multiple_channels.feature:26]: Here I use a new transformer `LocaleContext::castToItsLocaleName`. I had a problem that UI expects locale names in polish while in API I check the code.

File: src/Sylius/Behat/Context/Api/Shop/ChannelContext.php [hunk 2]
```diff
@@ -24,22 +25,29 @@ final class ChannelContext implements Context
 {
     private ApiClientInterface $client;
 
+    private ChannelContextSetterInterface $channelContextSetter;
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/ChannelContext.php:28]: do we use it somewhere?

File: src/Sylius/Behat/Context/Api/Shop/ChannelContext.php [hunk 3]
```diff
@@ -24,22 +25,29 @@ final class ChannelContext implements Context
...
+        $this->channelContextSetter = $channelContextSetter;
         $this->responseChecker = $responseChecker;
         $this->sharedStorage = $sharedStorage;
     }
 
     /**
      * @Given I am browsing channel :channel
+     * @When /^I (?:am browsing|start browsing|try to browse|browse) (that channel)$/
+     * @When /^I (?:am browsing|start browsing|try to browse|browse) (?:|the )("[^"]+" channel)$/
+     * @When /^I (?:am browsing|start browsing|try to browse|browse) (?:|the )(channel "[^"]+")$/
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/ChannelContext.php:50]: why do we need all of these steps?

File: src/Sylius/Behat/Context/Api/Shop/LocaleContext.php [hunk 4]
```diff
@@ -85,9 +90,23 @@ public function theLocaleWithCodeShouldNotBeAvailable(string $name, string $code
...
 
         Assert::true($this->responseChecker->hasValue($response, 'name', $name));
         Assert::true($this->responseChecker->hasValue($response, 'code', $code));
     }
+
+    /**
+     * @Then I should shop using the :localeCode locale
+     * @Then I should still shop using the :localeCode locale
+     */
+    public function iShouldShopUsingTheLocale(string $localeCode)
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/LocaleContext.php:103]:
```suggestion
    public function iShouldShopUsingTheLocale(string $localeCode): void
```

File: src/Sylius/Behat/Context/Api/Shop/LocaleContext.php [hunk 5]
```diff
@@ -85,9 +90,23 @@ public function theLocaleWithCodeShouldNotBeAvailable(string $name, string $code
...
-        $response = $this->client->getLastResponse();
+        $response = $this->localesClient->getLastResponse();
 
         Assert::true($this->responseChecker->hasValue($response, 'name', $name));
         Assert::true($this->responseChecker->hasValue($response, 'code', $code));
     }
+
+    /**
+     * @Then I should shop using the :localeCode locale
+     * @Then I should still shop using the :localeCode locale
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/LocaleContext.php:101]:
```suggestion
     * @Then I should( still) shop using the :localeCode locale
```
and line 100 will be redundant :)

File: src/Sylius/Behat/Context/Api/Shop/LocaleContext.php [hunk 6]
```diff
@@ -85,9 +90,23 @@ public function theLocaleWithCodeShouldNotBeAvailable(string $name, string $code
...
+     */
+    public function iShouldShopUsingTheLocale(string $localeCode)
+    {
+        $this->cartsClient->buildCreateRequest();
+
+        Assert::same(
+            $this->responseChecker->getValue($this->cartsClient->create(), 'localeCode'),
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Api/Shop/LocaleContext.php:108]: I know that we talked about it on audio but the default locale shouldn't be taken from the cart (we should get it from the channel).

File: src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php [hunk 7]
```diff
@@ -48,12 +48,12 @@ public function iShowHomepageWithTheLocale(string $localeCode): void
     }
 
     /**
-     * @Then I should shop using the :locale locale
-     * @Then I should still shop using the :locale locale
+     * @Then I should shop using the :itsLocaleName locale
+     * @Then I should still shop using the :itsLocaleName locale
      */
-    public function iShouldShopUsingTheLocale($locale)
+    public function iShouldShopUsingTheLocale(string $itsLocaleName)
```
Comment (AdamKasp) [inline src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php:54]:
```suggestion
    public function iShouldShopUsingTheLocale(string $LocaleName): void
```
IMO `localeName` will be enough

Comment (coldic3) [reply src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php:54]: Doesn't work that way :confused: I think the variable name has to be the same as defined in the transformer.

File: features/locale/handling_different_locales_on_multiple_channels.feature [hunk 8]
```diff
@@ -1,33 +1,33 @@
...
         And that channel allows to shop using "Polish (Poland)" and "Norwegian (Norway)" locales
         And it uses the "Polish (Poland)" locale by default
 
     @ui
     Scenario: Showing locales only from the current channel
         When I browse the "Mobile" channel
-        Then I should shop using the "polski (Polska)" locale
+        Then I should shop using the "Polish (Poland)" locale
         And I should be able to shop using the "norweski (Norwegia)" locale
         And I should not be able to shop using the "angielski (Stany Zjednoczone)" locale
```
Comment (GSadee) [inline features/locale/handling_different_locales_on_multiple_channels.feature:20]: These ones should also be in English to make steps consistent

File: src/Sylius/Behat/Context/Api/Shop/LocaleContext.php [hunk 9]
```diff
@@ -17,19 +17,24 @@
 use Sylius\Behat\Client\ApiClientInterface;
 use Sylius\Behat\Client\ResponseCheckerInterface;
 use Sylius\Component\Locale\Model\LocaleInterface;
+use Symfony\Component\Intl\Locales;
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Api/Shop/LocaleContext.php:20]: Do you use it in this context?

File: src/Sylius/Behat/Context/Transform/LocaleContext.php [hunk 10]
```diff
@@ -42,6 +42,16 @@ public function castToLocaleCode(string $localeName): string
...
     }
 
+    /**
+     * @Transform :itsLocaleName
+     */
+    public function castToItsLocaleName(string $localeName): string
+    {
+        $code = $this->localeNameConverter->convertNameToCode($localeName);
+
+        return $this->localeNameConverter->convertCodeToName($code, $code);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Transform/LocaleContext.php:52]:
```suggestion
        $localeCode = $this->localeNameConverter->convertNameToCode($localeName);

        return $this->localeNameConverter->convertCodeToName($localeCode, $localeCode);
```

File: features/locale/handling_different_locales_on_multiple_channels.feature (hunk 8 above)
Comment (coldic3) [reply features/locale/handling_different_locales_on_multiple_channels.feature:20]: The tricky part here is that we have a non-polish locale with polish translation. So then I would have to do something like `:localNameInPolishLocale` :thinking:

File: src/Sylius/Behat/Context/Api/Shop/LocaleContext.php (hunk 6 above)
Comment (coldic3) [reply src/Sylius/Behat/Context/Api/Shop/LocaleContext.php:106]: Indeed, that makes sense now for me, so I'll keep it as it is.

File: features/locale/handling_different_locales_on_multiple_channels.feature (hunk 8 above)
Comment (coldic3) [reply features/locale/handling_different_locales_on_multiple_channels.feature]: I did `:localeNameInCurrentLocale` so `:localeNameInItsLocale` is now redundant :slightly_smiling_face:

File: features/locale/handling_different_locales_on_multiple_channels.feature (hunk 1 above)
Comment (coldic3) [reply features/locale/handling_different_locales_on_multiple_channels.feature:26]: It is handled now by `:localeNameInCurrentLocale`.

File: src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php [hunk 11]
```diff
@@ -15,54 +15,66 @@
...
-     * @When I switch to the :name locale
-     * @When I change my locale to :name
+     * @When I show homepage with the locale :locale
      */
-    public function iSwitchTheLocaleToTheLocale(string $name): void
+    public function iShowHomepageWithTheLocale(LocaleInterface $locale): void
     {
-        $this->homePage->open();
-        $this->homePage->switchLocale($name);
+        $this->homePage->tryToOpen(['_locale' => $locale->getCode()]);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php:50]: Why did you change that? I would be for leaving as it was
```suggestion
        $this->homePage-> switchLocale($locale->getName());
```

File: src/Sylius/Behat/Resources/config/suites/api/locale/locales.yaml [hunk 12]
```diff
@@ -15,6 +15,8 @@ default:
                 - sylius.behat.context.setup.channel
                 - sylius.behat.context.setup.locale
 
+                - sylius.behat.context.api.shop.channel
+                - sylius.behat.context.api.shop.cart
```
Comment (GSadee) [inline src/Sylius/Behat/Resources/config/suites/api/locale/locales.yaml:19]: It can be in alphabetical order

**Review by GSadee — APPROVED** [review]

File: src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php (hunk 11 above)
Comment (coldic3) [reply src/Sylius/Behat/Context/Ui/Shop/LocaleContext.php:50]: I didn't, GitHub shows diff it in a strange way. This change you highlighted is for the `iShowHomepageWithTheLocale`.

**Comment by GSadee** [comment]
Thanks, Kevin! :1st_place_medal:
